data/dataset_neo.py

```diff
@@ -0,0 +1,338 @@
+        self.split_type = split_type
+
+    def __len__(self) -> int:
+        return len(self.users)
+
+    def __getitem__(self, idx: int) -> Union[Data, HeteroData]:
+        """Create Edges"""
+        all_edges = self.graph[Constants.edge_key].edge_index
+
+        positive_article_indices = t.as_tensor(
+            self.users[idx], dtype=t.long
+        )  # all the positive target indices for the current user
+        positive_article_edges = create_edges_from_target_indices(
+            idx, positive_article_indices
+        )
+
+        # Sample positive edges from subgraph (amount defined in config.positive_edges_ratio)
+        samp_cut = max(
+            1,
+            math.floor(
+                len(positive_article_indices) * self.config.positive_edges_ratio
+            ),
+        )
+
+        if self.randomization:
+            random_integers = t.randint(
+                low=0, high=len(positive_article_indices), size=(samp_cut,)
+            )
+        else:
+            random_integers = t.tensor(
+                [
+                    t.min(positive_article_indices, dim=0)[1].item(),
+                    t.max(positive_article_indices, dim=0)[1].item(),
+                ]
+            )
+
+        sampled_positive_article_indices = positive_article_indices[random_integers]
+        sampled_positive_article_edges = create_edges_from_target_indices(
+            idx, sampled_positive_article_indices
+        )
+
+        num_sampled_pos_edges = sampled_positive_article_indices.shape[0]
+        if num_sampled_pos_edges <= 1:
+            negative_edges_ratio = self.config.k - 1
+        else:
+            negative_edges_ratio = self.config.negative_edges_ratio
+
+        if self.train:
+            # Randomly select from the whole graph
+            sampled_negative_article_edges = create_edges_from_target_indices(
+                idx,
+                get_negative_edges_random(
+                    subgraph_edges_to_filter=sampled_positive_article_indices,
+                    all_edges=all_edges,
+                    num_negative_edges=int(
+                        negative_edges_ratio * num_sampled_pos_edges
+                    ),
+                    randomization=self.randomization,
+                ),
+            )
+        else:
+            assert self.matchers is not None, "Must provide matchers for test"
+            # Select according to a heuristic (eg.: lightgcn scores)
+            candidates = t.cat(
+                [matcher.get_matches(idx) for matcher in self.matchers],
+                dim=0,
+            ).unique()
+            # but never add positive edges
+            sampled_negative_article_edges = create_edges_from_target_indices(
+                idx,
+                only_items_with_count_one(
+                    t.cat([candidates, positive_article_indices], dim=0)
+                ),
+            )
+
+        # n_hop_edges_old = fetch_n_hop_neighbourhood(
```

Comment:
I think this is not needed anymore?

Reply:
I left it in so that we can compare the two results one after another. But it's just for sanity check, we can delete it otherwise.